src/counts.jl

```diff
@@ -42,15 +42,21 @@ function addcounts!(r::AbstractArray, x::IntegerArray, levels::IntUnitRange)
 end
 
 function addcounts!(r::AbstractArray, x::IntegerArray, levels::IntUnitRange, wv::AbstractWeights)
-    k = length(levels)
-    length(r) == k || throw(DimensionMismatch())
+    # add wv weighted counts of integers from x that fall within levels to r
+
+    length(x) == length(wv) ||
+        throw(DimensionMismatch("x and wv must have the same length, got $(length(x)) and $(length(wv))"))
 
-    m0 = levels[1]
-    m1 = levels[end]
+    xv = vec(x) # discard shape because weights() discards shape
+
+    checkbounds(r, axes(levels)...)
+
+    m0 = first(levels)
+    m1 = last(levels)
     b = m0 - 1
 
-    @inbounds for i in 1 : length(x)
-        xi = x[i]
+    @inbounds for i in eachindex(xv, wv)
```

Comment:
this will fail if `x` is `OffsetArray`:
```
julia> eachindex(vec(OffsetArray(1:4, 3)), Weights(1:4))
ERROR: DimensionMismatch
```

Reply:
Previously, addcounts of an OffsetArray and a non-offset Weights resulted in a segfault. In general, I am opposed to silently supporting operations with mismatched indices, and in this particular case, because it is not a regression to throw a dimension mismatch instead of segfaulting, I'd like to call that error out of scope.

Reply:
OK - I thought you wanted to fix OffsetArrays.jl issues in this PR, but agreed it can go to a separate PR since what you propose now is safe.

However note that in:
> `xv = vec(x)` # discard shape because weights() discards shape

you already decide on the interface you mention - choosing to support "position in sequence" interpretation (if `x` is higher dimensional you implicitly perform this matching)

Reply:
That choice was forced to avoid a regression on `addcounts(zeros(Int, 5), rand(1:5, 3, 3), Weights(rand(9)))`.

It is also an implicit conversion to linear indexing, which is somewhat different than an implicit alignment of mismatched indices.

> OK - I thought you wanted to fix OffsetArrays.jl issues in this PR

I don't see this as an issue:

```julia
julia> w = Weights([4,2,6]);

julia> ow = Weights(OffsetVector([4,2,6], 2));

julia> x = [3,1,2];

julia> ox = OffsetVector([3,1,2], 2);

julia> addcounts!(zeros(Int, 3), x, 1:3, w)
3-element Vector{Int64}:
 2
 6
 4

julia> addcounts!(zeros(Int, 3), ox, 1:3, ow)
3-element Vector{Int64}:
 2
 6
 4

julia> addcounts!(zeros(Int, 3), ox, 1:3, w)
ERROR: DimensionMismatch: all inputs to eachindex must have the same axes, got (OffsetArrays.IdOffsetRange(values=3:5, indices=3:5),) and (Base.OneTo(3),)
Stacktrace:
 [1] throw_eachindex_mismatch_indices(::IndexCartesian, ::Tuple{OffsetArrays.IdOffsetRange{Int64, Base.OneTo{Int64}}}, ::Vararg{Any})
   @ Base ./abstractarray.jl:292
 [2] eachindex
   @ ./multidimensional.jl:388 [inlined]
 [3] eachindex
   @ ./abstractarray.jl:334 [inlined]
 [4] addcounts!(r::Vector{Int64}, x::OffsetVector{Int64, Vector{Int64}}, levels::UnitRange{Int64}, wv::Weights{Int64, Int64, Vector{Int64}})
   @ StatsBase ~/.julia/dev/StatsBase/src/counts.jl:58
 [5] top-level scope
   @ REPL[24]:1

julia> addcounts!(zeros(Int, 3), x, 1:3, ow)
ERROR: DimensionMismatch: all inputs to eachindex must have the same axes, got (Base.OneTo(3),) and (OffsetArrays.IdOffsetRange(values=3:5, indices=3:5),)
Stacktrace:
 [1] throw_eachindex_mismatch_indices(::IndexCartesian, ::Tuple{Base.OneTo{Int64}}, ::Vararg{Any})
   @ Base ./abstractarray.jl:292
 [2] eachindex
   @ ./multidimensional.jl:388 [inlined]
 [3] eachindex
   @ ./abstractarray.jl:334 [inlined]
 [4] addcounts!(r::Vector{Int64}, x::Vector{Int64}, levels::UnitRange{Int64}, wv::Weights{Int64, Int64, OffsetVector{Int64, Vector{Int64}}})
   @ StatsBase ~/.julia/dev/StatsBase/src/counts.jl:58
 [5] top-level scope
   @ REPL[25]:1

julia> 

```

Reply:
To allow different index styles (but same length) we could do:

```for (xi, wi) in zip(xv, wv)```

Reply:
`zip` is too tolerant, it even allows inputs of different sizes and silently ignores extra elements in the other argument. As mentioned in the main discussion, we should probably be strict about having matching indices or things could be really confusing.